### PR TITLE
DSL query language #8936

### PR DESCRIPTION
**Pull request by vbradnitski**

### CODE REVIEW

File: modules/lib/lib-node/src/test/resources/test/FindNodesByQueryHandlerTest.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+var nodeLib = require('/lib/xp/node');
+var assert = require('/lib/xp/testing');
+
+var repo = nodeLib.connect({
+    repoId: 'com.enonic.cms.default',
```
Comment (alansemenov) [inline modules/lib/lib-node/src/test/resources/test/FindNodesByQueryHandlerTest.js:5]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
    repoId: "com.enonic.cms.default",
```

File: modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+            'page': {},
+            'attachments': {},
+            'publish': {},
+            'workflow': {
+                'state': 'READY',
+                'checks': {}
+            }
+        },
+        {
+            '_id': 'id3',
```
Comment (alansemenov) [inline modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js:53]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
            "_id": 'id3',
```

File: modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        'price_ranges': {
+            'buckets': [
+                {
+                    'key': 'a',
+                    'docCount': 2,
+                    'to': 50
+                },
+                {
+                    'key': 'b',
+                    'docCount': 4,
```
Comment (alansemenov) [inline modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js:113]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
                    "docCount": 4,
```

File: modules/lib/lib-node/src/main/resources/lib/xp/examples/node/query-dsl.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            }
+        }
+    }
+});
+
+log.info('Found ' + result.total + ' number of contents');
+
+for (var i = 0; i < result.hits.length; i++) {
+    var node = result.hits[parseInt(i)];
+    log.info('Node ' + node.id + ' found');
```
Comment (alansemenov) [inline modules/lib/lib-node/src/main/resources/lib/xp/examples/node/query-dsl.js:33]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
    log.info("Node " + node.id + ' found');
```

File: modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js [hunk 5]
```diff
@@ -0,0 +1,299 @@
+var assert = require('/lib/xp/testing.js');
+var content = require('/lib/xp/content.js');
+
+var expectedJson = {
+    'total': 20,
+    'count': 3,
+    'hits': [
+        {
+            '_id': 'id1',
+            '_name': 'name1',
```
Comment (alansemenov) [inline modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js:10]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
            "_name": 'name1',
```

File: modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+            'modifier': 'user:system:admin',
+            'createdTime': '1970-01-01T00:00:00Z',
+            'modifiedTime': '1970-01-01T00:00:00Z',
+            'type': 'base:unstructured',
+            'displayName': 'My Content 1',
+            'hasChildren': false,
+            'valid': false,
+            'data': {},
+            'x': {},
+            'page': {},
```
Comment (alansemenov) [inline modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js:22]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
            "page": {},
```

File: modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+                    'key': 'female',
+                    'docCount': 12
+                }
+            ]
+        },
+        'by_month': {
+            'buckets': [
+                {
+                    'key': '2014-01',
+                    'docCount': 8
```
Comment (alansemenov) [inline modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js:92]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
                    "docCount": 8
```

File: modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+                    'key': 'male',
+                    'docCount': 10
+                },
+                {
+                    'key': 'female',
+                    'docCount': 12
+                }
+            ]
+        },
+        'by_month': {
```
Comment (alansemenov) [inline modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js:88]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
        "by_month": {
```

File: modules/lib/lib-node/src/test/resources/test/FindNodesByQueryHandlerTest.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        repo.query({
+            'start': 0,
+            'count': 100,
+            'query':
+                [{
+                    boolean: {
+                        must: [{
+                            term: {
+                                name: 'type',
+                                value: 'article'
```
Comment (alansemenov) [inline modules/lib/lib-node/src/test/resources/test/FindNodesByQueryHandlerTest.js:21]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
                                value: "article"
```

File: modules/lib/lib-node/src/test/resources/test/FindNodesByQueryHandlerTest.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
+var nodeLib = require('/lib/xp/node');
+var assert = require('/lib/xp/testing');
+
+var repo = nodeLib.connect({
+    repoId: 'com.enonic.cms.default',
+    branch: 'master'
```
Comment (alansemenov) [inline modules/lib/lib-node/src/test/resources/test/FindNodesByQueryHandlerTest.js:6]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
    branch: "master"
```

File: modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+            'type': 'base:unstructured',
+            'displayName': 'My Content 1',
+            'hasChildren': false,
+            'valid': false,
+            'data': {},
+            'x': {},
+            'page': {},
+            'attachments': {},
+            'publish': {},
+            'workflow': {
```
Comment (alansemenov) [inline modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js:25]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
            "workflow": {
```

File: modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+                    'key': '2014-02',
+                    'docCount': 10
+                },
+                {
+                    'key': '2014-03',
+                    'docCount': 12
+                }
+            ]
+        },
+        'price_ranges': {
```
Comment (alansemenov) [inline modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js:104]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=8692790)
```suggestion
        "price_ranges": {
```

File: modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/InQueryBuilder.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    InQueryBuilder( final PropertySet expression )
+    {
+        super( expression );
+
+        this.values = getObjects( "values" );
+    }
+
+    public QueryBuilder create()
+    {
+        if ( values == null || values.isEmpty() )
```
Comment (vbradnitski) [inline modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/InQueryBuilder.java:29]: nullToEmpty( value ).isBlank() or empty

File: modules/core/core-repo/src/test/resources/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/fulltext/result/boosted.json [hunk 14]
```diff
@@ -0,0 +1,13 @@
+{
+  "simple_query_string": {
+    "query": "fisk",
```
Comment (rymsha) [inline modules/core/core-repo/src/test/resources/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/fulltext/result/boosted.json:3]: document special chars

File: modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/ExpressionQueryBuilder.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+            case "date":
+                return ValueTypes.LOCAL_DATE.convert( value );
+            case "dateTime":
+                try
+                {
+                    return ValueTypes.DATE_TIME.convert( value );
+                }
+                catch ( ValueTypeException e )
+                {
+                    return ValueTypes.LOCAL_DATE_TIME.convert( value ).toInstant( ZoneOffset.UTC );
```
Comment (rymsha) [inline modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/ExpressionQueryBuilder.java:50]: I'm not very happy about catch in normal flow

File: modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/SimpleStringQueryBuilder.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.SimpleQueryStringBuilder;
+
+import com.enonic.xp.data.PropertySet;
+import com.enonic.xp.repo.impl.elasticsearch.query.translator.factory.function.WeightedQueryFieldNames;
+import com.enonic.xp.repo.impl.elasticsearch.query.translator.resolver.SearchQueryFieldNameResolver;
+
+import static com.google.common.base.Strings.nullToEmpty;
+
+abstract class SimpleStringQueryBuilder
```
Comment (rymsha) [inline modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/SimpleStringQueryBuilder.java:14]: Maybe `SimpleQueryStringBuilder` ?

File: modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/SimpleStringQueryBuilder.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+    protected final SearchQueryFieldNameResolver nameResolver;
+
+    SimpleStringQueryBuilder( final PropertySet expression )
+    {
+        super( expression );
+
+        this.query = getString( "query" );
+        this.operator = getString( "operator" );
+        final String field = getString( "field" );
```
Comment (rymsha) [inline modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/SimpleStringQueryBuilder.java:31]: It is confusing to support both singular and plural form `fields`. What is the idea here?

File: modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/function/WeightedQueryFieldNames.java [hunk 18]
```diff
@@ -12,23 +13,17 @@
...
     }
 
-    public ImmutableList<WeightedQueryFieldName> getWeightedQueryFieldNames()
-    {
-        return weightedQueryFieldNames;
-    }
-
-    @Override
-    public Iterator<WeightedQueryFieldName> iterator()
+    public static WeightedQueryFieldNames from( final Collection<String> weightedQueryFieldNames )
```
Comment (rymsha) [inline modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/function/WeightedQueryFieldNames.java:23]: Maybe we should not support weighted names in DSL? Instead give them weights?

File: modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/ExpressionQueryBuilder.java [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+        switch ( type )
+        {
+            case "time":
+                return ValueTypes.LOCAL_TIME.convert( value );
+            case "date":
+                return ValueTypes.LOCAL_DATE.convert( value );
+            case "dateTime":
+                return ValueTypes.DATE_TIME.convert( value );
+            case "localDateTime":
+                return ValueTypes.LOCAL_DATE_TIME.convert( value ).toInstant( ZoneOffset.UTC );
```
Comment (rymsha) [inline modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/elasticsearch/query/translator/factory/dsl/ExpressionQueryBuilder.java:45]: why we convert to UTC only localDateTime, and not localDate ? They should be handled equally.

**Review by rymsha — APPROVED** [review]

**Review by anatol-sialitski — APPROVED** [review]

**Pull request merged**